src/auth/aws.js

```diff
@@ -45,6 +50,86 @@ export function createCanonicalRequest(request) {
+ * @returns 
+ */
+export function createStringToSign(canonicalRequest) {
+    if (!canonicalRequest) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof canonicalRequest !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    const timeStampISO8601Format = _.getAWSTimestamp();
+    const scope = _.createScope();
+    const request = hex(sha256(canonicalRequest).toString());
+
+    return `${AUTH_IDENTIFIER_HEADER}\n${timeStampISO8601Format}\n${scope}\n${request}`;
+}
+
+/**
+ * Returns Signing key
+ * 
+ * @export
+ * @param {any} awsSecretKey 
+ * @param {any} awsRegion 
+ * @returns {String}
+ */
+export function createSigningKey(awsSecretKey, awsRegion) {
+    if (!awsSecretKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsSecretKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!awsRegion) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsRegion !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return _.toHmacSHA256(_.toHmacSHA256(_.toHmacSHA256(_.toHmacSHA256(`AWS4${awsSecretKey}`, _.getShortDate()), awsRegion), 's3'), 'aws4_request');
+}
+
+/**
+ * Returns signature
+ * 
+ * @export
+ * @param {any} signingKey 
+ * @param {any} stringToSign 
+ * @returns {String}
+ */
+export function createSignature(signingKey, stringToSign) {
+    if (!signingKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof signingKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!stringToSign) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof stringToSign !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return _.toHmacSHA256(signingKey, stringToSign);
+}
+
+
+/**
+ * Returns Authorization Header
+ * 
+ * @export
+ * @param {any} awsAccessKey 
+ * @param {any} awsRegion 
+ * @param {any} signature 
+ * @returns {String}
+ */
+export function createAuthorizationHeader(awsAccessKey, awsRegion, signature) {
+    if (!awsAccessKey) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsAccessKey !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!awsRegion) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof awsRegion !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+    if (!signature) throw Error(ERROR_PARAM_REQUIRED);
+    if (typeof signature !== 'string') throw Error(ERROR_PARAM_TYPE_IS_NOT_STRING);
+
+    return `${AUTH_IDENTIFIER_HEADER} Credential=${awsAccessKey}/${_.getShortDate()}/${awsRegion}/s3/aws4_request,SignedHeaders=host;range;x-amz-content-sha256;x-amz-date,Signature=${signature}`;
+}
+
+export function ammendRequest(request) {
+    const canonicalRequest = createCanonicalRequest(request);
+    const stringToSign = createStringToSign(canonicalRequest);
+    const signingKey = createSigningKey(AWS_SECRET_ACCESS_KEY, AWS_REGION);
+    const signature = createSignature(signingKey, stringToSign);
+    createAuthorizationHeader(AWS_ACCESS_KEY_ID, AWS_REGION, signature);
+    // TODO: Ammend the request
```

Comment:
I don't actually think this is amending a request - it's creating a completely new request using the data from the old one...

Not sure that's just semantics, think it ought to actually  be how we code it as well...?